Bottom: 4cc44fca8f0014d48b531c9cc0834c9618490654
Top:    384e5e94782f25233af9c7a4b748775beb5e4fca
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-08-02 13:19:02 +0800

tuntap: flush xdp map in the caller

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 3f23e2c..7951012 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1626,9 +1626,10 @@ static u32 tun_do_xdp(struct tun_struct *tun,
 	switch (act) {
 	case XDP_REDIRECT:
 		*err = xdp_do_redirect(tun->dev, xdp, xdp_prog);
-		xdp_do_flush_map();
-		if (*err)
+		if (*err) {
+			act = XDP_DROP;
 			break;
+		}
 		goto out;
 	case XDP_TX:
 		*err = tun_xdp_tx(tun->dev, xdp);
@@ -1715,6 +1716,9 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 		act = tun_do_xdp(tun, tfile, xdp_prog, &xdp, &err);
 		if (err)
 			goto err_xdp;
+
+		if (act == XDP_REDIRECT)
+			xdp_do_flush_map();
 		if (act != XDP_PASS)
 			goto out;
